### ISSUE TITLE
FOUC in Production NextJS apps

### ISSUE THREAD

**Issue by megancooper** [mantinedev/mantine#1214]
### What package has an issue

@mantine/core

**_Note: This issue only occurs in a production nextjs build. It is not noticeable while running the app in development_**

### Describe the bug

Hello! I've noticed a quick jerk/blink that seems to happen before styles load. (See the gif below for visual). I'm not sure what's causing this behavior, but I wanted to know if I was doing something wrong in my implementation. 

I'm currently using vercel to deploy the site and using domain forwarding. Could this be the culprit?

Thanks for any help!

Here is my `_document` file:
```
import Document, {
  Html, Head, Main, NextScript,
} from 'next/document';
import {createGetInitialProps} from '@mantine/next';

const getInitialProps = createGetInitialProps();

export default class _Document extends Document {
  static getInitialProps = getInitialProps;

  render() {
    return (
      <Html lang="en">
        <Head>
          <link rel="shortcut icon" href="/static/favicon.ico" />
        </Head>
        <body className="light">
          <Main />
          <NextScript />
        </body>
      </Html>
    );
  }
}
```

And the `_app` file:
```
import Head from 'next/head';
import {AppProps} from 'next/app';
import {MantineProvider} from '@mantine/core';
import {SubscribeContextProvider} from '../contexts/SubscribeContext';
import {FeedbackContextProvider} from '../contexts/FeedbackContext';
import Navigation from '../components/Navigation';
import Footer from '../components/Footer';

const App = ({Component, pageProps}: AppProps) => (
  <>
    <Head>
      <title>Melo</title>
    </Head>
    <MantineProvider
      withGlobalStyles
      withNormalizeCSS
      theme={{
        colorScheme: 'light',
      }}
    >
      <SubscribeContextProvider>
        <FeedbackContextProvider>
          <Navigation />
          <Component {...pageProps} />
          <Footer />
        </FeedbackContextProvider>
      </SubscribeContextProvider>
    </MantineProvider>
  </>
);

export default App;
```

### **Here is a gif showing the behavior:**

![page_load](https://user-images.githubusercontent.com/8683055/162582981-5e5b1192-d36e-41f3-aa8a-f9b406905fe4.gif)

### In which browser did the problem occur

chrome

### Do you know how to fix the issue

No

### Are you willing to participate in fixing this issue and create a pull request with the fix

Yes

**Comment by rtivital** [comment]
You have incorrect ssr logic. It is always hard to debug, so we usually just advise to use next template, it is known to not have any issues. https://github.com/mantinedev/mantine-next-template

**Comment by megancooper** [comment]
@rtivital Thanks, I'll check this out! Is there something specific that's wrong? I'd just like to know so that as I add to the template I don't encounter weird rendering behavior. 🙏🏽

**Comment by rtivital** [comment]
No, ssr setup is very straightforward it requires only modifications to _document.tsx. Note that we do not support experimental react and next.js versions.

**Comment by megancooper** [comment]
@rtivital Ah thanks 🙏🏽  Apologies I was referencing my own implementation not the template. It sounds like I shouldn't be modifying the `_document` file the way that I am currently. I'll make changes based on the template and try them out 👍🏽

**Comment by megancooper** [comment]
I made changes based on the template, but am still seeing the blinking issue. Here is a link to the code for my site:
https://github.com/megancooper/personal-website
Is there something wrong with how I've implemented anything? Thanks 🙏🏽

**Comment by rtivital** [comment]
Sorry, I cannot help with any further investigation, I've created a new project following the guide from documentation, and it does not have any issues – https://github.com/rtivital/mantine-1214. I can only suggest that you use it or mantine template as a base.

**Comment by megancooper** [comment]
Thank you, it could be related to the free-tier vercel service I'm using...will investigate further 👍🏽

**Comment by megancooper** [comment]
Found another stackoverflow post tracking this issue for anyone that lands on this thread:
https://stackoverflow.com/questions/71400494/flash-of-unstyled-content-fouc-for-nextjs-using-mantine

**Comment by ilomon10** [comment]
@megancooper how do you solve this problem?

any updates?

**Comment by megancooper** [comment]
@ilomon10 Still ongoing, was planning on taking a deeper look into the issue this week

**Comment by simonhochrein** [comment]
> @rtivital I just tried the repo you created [here](https://github.com/rtivital/mantine-1214) and it show the same exact problem people are describing here.
> 
> The only difference I see is that I use NPM and you Yarn.
> 
> The problem I suppose is cause by this `<style data-emotion="mantine "></style>` being empty.
> 
> I've done some more digging and if it's installed with Yarn everything works fine, no flash, if instead it's being installed with NPM it somehow breaks.

Can confirm it works with yarn, but it is broken with pnpm. I have not tried with vanilla npm.

**Comment by megancooper** [comment]
@simonhochrein @ilomon10 
Not sure if this will fix the issue for you all, but I was able to resolve my issue my updated my @mantine/next package. I was using `@mantine/next@4.0.9` and upgraded to `@mantine/next@4.1.3` (+ `@mantine/core@4.1.3`)  which resolved the issue

**Comment by rtivital** [comment]
Seems like npm/pnpm just randomly rearranges `@emotion/` packages versions, updating versions does not usually help

**Comment by erdimaden** [comment]
Thanks for sharing this. I'm using `4.2.6` and the same issue does exist. 

any updates on this issue?

> @simonhochrein @ilomon10 Not sure if this will fix the issue for you all, but I was able to resolve my issue by updating the `@mantine/next` package. I was using `@mantine/next@4.0.9` and upgraded to `@mantine/next@4.1.3` (+ `@mantine/core@4.1.3`) which resolved the issue
> 
> Edit: FWIW I'm using yarn

**Comment by NathanLandsX** [comment]
Same. I'm using Nextjs/Mantine/tRPC.

**Comment by ilomon10** [comment]
https://github.com/mantinedev/mantine/issues/1148#issuecomment-1109382264
In my case, i just fixed with removing `emotionOptions` attribute from `MantineProvider`

**Comment by rtivital** [comment]
From FAQ:
**Next.js SSR does not work/while page is loading there are no styles/styles do not work without JavaScript**
You haven't configured SSR correctly. To fix that use template (https://github.com/mantinedev/mantine-next-template) or:
- Follow the Next.js getting started guide – https://mantine.dev/theming/next/
- Delete .next folder to reset cache
- If that does not help try using yarn instead of npm to install dependencies
- If you are using custom emotionOptions make sure that key: 'mantine'
- Note that Next.js of version 12.1.5 has ssr issues with emotion, you should install other version

**Comment by mcamendoza1** [comment]
Downgrading from next@12.1.6 to 12.1.5 fixed the issue

**Comment by jason-dark** [comment]
> Downgrading from next@12.1.6 to 12.1.5 fixed the issue

+1
Had followed all the instructions above, Next.js setup mirroring the docs and FOUC persisted.
Downgrading to next@12.1.5 solved it for me too.

**Comment by klen** [comment]
Have same problem with `next@12.1.6`. Downgrading to `12.1.5` solve it. But, upgrading to `12.1.7-canary.29` also did the job.

**Comment by ISnackable** [comment]
> Have same problem with next@12.1.6. Downgrading to 12.1.5 solve it. But, upgrading to 12.1.7-canary.29 also did the job.

Thank you! This issue was driving me nuts. Upgrading NextJS to `12.1.7-canary.31` fixed it for me.
```sh
yarn add next@v12.1.7-canary.31
```

**Comment by fz6m** [comment]
I fixed it by upgrading nextjs to `12.1.7-canary.33` 😊

**Comment by Nawbc** [comment]
> I fixed it by upgrading nextjs to `12.1.7-canary.33` blush

I had a try, but not for me.

**Comment by Nawbc** [comment]
> From FAQ: **Next.js SSR does not work/while page is loading there are no styles/styles do not work without JavaScript** You haven't configured SSR correctly. To fix that use template (https://github.com/mantinedev/mantine-next-template) or:
> 
> * Follow the Next.js getting started guide – https://mantine.dev/theming/next/
> * Delete .next folder to reset cache
> * If that does not help try using yarn instead of npm to install dependencies
> * If you are using custom emotionOptions make sure that key: 'mantine'
> * Note that Next.js of version 12.1.5 has ssr issues with emotion, you should install other version

I have tried all but still has the problem.  next:12.1.5, 12.1.6, 12.1.7-canary.35    https://letscollab.deskbtm.com

**Comment by theskollie** [comment]
Dealing with the same issue. I've also tried next:12.1.5, 12.1.6, 12.1.7-canary.35.

**Comment by mcamendoza1** [comment]
@theskollie you can try the latest version of next:12.2.0 looks fine on my side.

**Comment by theskollie** [comment]
@mcamendoza1 Thanks - Just updated to next": "^12.2.0" and still experiencing FOUC on Firefox.

**Comment by mcamendoza1** [comment]
> @mcamendoza1 Thanks - Just updated to next": "^12.2.0" and still experiencing FOUC on Firefox.

haven't tested it firefox, can you check it in chrome?

**Comment by zwying0814** [comment]
I removed the emotionOptions setting and it seems to be working fine, but why?

**Comment by ilomon10** [comment]
@zwying0814 
> * If you are using custom emotionOptions make sure that key: 'mantine'
> * Note that Next.js of version 12.1.5 has ssr issues with emotion, you should install other version

**Comment by NeoHuncho** [comment]
I am still getting the issue with v5. I am using the T3 stack ( Tailwind and Next)

**Comment by rtivital** [comment]
Follow Next.js getting started guide, use one of the templates as reference

**Comment by based64-eth** [comment]
> https://github.com/mantinedev/mantine-next-template

The @emotion/server dependency is not documented.

**Comment by rtivital** [comment]
Emotion server is not a mantine package. You can find documentation for it on emotion website.
